### DISCUSSION TITLE
The illuminated point cloud is not on the side of the light source

### DISCUSSION THREAD

**Discussion by mingkangzh** [vsg-dev/VulkanSceneGraph#1055] [Q&A] [answered]
Hello.
In my scene, there is a sphere, a box, and two point clouds with opposite normals.

I found that point light sources can work properly on solid objects, but there is a slight issue with point clouds.

I want the light source and the normal of the point cloud to be on the same side, thus illuminating the point cloud.

But in reality, the point cloud that is always not on the side of the light source is illuminated.

In this scene, there are two point clouds with opposite normals. Normally, there would be a point cloud illuminated on the same side as the light source, but the expected result does not occur. The illuminated point cloud is always not on the side of the light source.

Do you know where the problem is? Is this normal?

Although one point cloud is bright, it does not have the color of a purple light source

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/5a9cbed0-15bf-4676-943d-ac25f240097c)

There is a purple point cloud above, but the illuminated point cloud surface and light source are not on the same side

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/333c8b99-5ad6-4f7e-8513-efa6186b41cf)

**Comment by mingkangzh** [comment]
The green box represents the location of the point light source

**Comment by mingkangzh** [comment]
In the case of the first image, the camera is facing the point cloud, and the surface of the point cloud is white. As the camera slowly tilts to view the point cloud, it gradually turns purple。

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/af831254-3329-4d3c-95c9-e7f477903fc4)

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/ef5db48c-ed39-4b89-91ab-6c72d2ab578e)

In the case of the first image, the light source is in front of the point cloud, the camera is facing the point cloud, and the surface of the point cloud is white. As the camera slowly tilts to view the point cloud, it gradually turns purple.
I think that since the light source is in front of the point cloud, it should also be purple when the camera is facing the point cloud. But it's still white

**Comment by robertosfield** [comment]
You haven't explain how the scene is constructed, or provided any details on the shaders used so there isn't anything we can suggest at this point.  The best we can do is ask questions.

1) How are you creating the point clouds and shape geometry?
2) What shaders or ShadersSet are you using?
3) Are you using the VSG's built in lights like vsg::Light?

**Reply by mingkangzh** [reply]
1. The point cloud directly creates a vertex array through points, specifying the instance normal. Shape geometry is created through a VSG builder.
2. Used brick_ ert and brick_phong_frag  ShaderSet .
3. Use VSG's built-in ambient light, directional light, and point light.

Thank you for your reminder. I will try to test whether the same problem will occur in vsgPoints.

**Reply by mingkangzh** [reply]
When the light source is in front of the object and the camera is facing it, the sphere turns purple, and the Box does not turn purple either. Normally, the Box should also be purple. Is this the reason for the shader?

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/13277b28-5691-4417-aab4-1122c0fa3c62)

**Reply by mingkangzh** [reply]
Sorry, my point cloud was too small before, so I thought there was an error.

Now I have loaded a large, white point cloud, and the white surface on the same side as the light source has achieved the desired effect.
The back is black, possibly due to the purple color assigned to the points when calculating the shader. The back is not completely black, and the parts illuminated by the light source are colored.

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/d76fa4af-1487-4cc0-94ef-7c923556b80d)

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/13cce191-1000-4fc1-b482-cfe23330fd8b)

However, when the light source is in front of the point cloud and the camera is facing the point cloud, the point cloud still does not change to the color of the light source.

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/630fdaee-5314-4ae8-8e36-de0c4fa138ef)

**Reply by robertosfield** [reply, marked as answer]
Perhaps it's just that the specular highlight is not what you want.

As general comment if some points are lit and others black then it'll usually be just done to normals pointing towards the light resulting in lit points, while normals pointing away results in ones only light be an ambient light if you have one assigned.

In the vsgPoints/data/shaders/brick_phong.frag that vsgPoints uses by default has a  VSG_TWO_SIDED_LIGHTING #define that you can assign to switch on the flipping of normal's that are facing away so both sides are lit.  You should be able to assign your own ShaderSet and enable this by default, though in the future it would be nice to have the ability to inherit the defines into the scene graph creation code so they can pass these defines on the GraphicsPipelineConfigurator when it's building the scene graph.

**Reply by mingkangzh** [reply]
Yes, point clouds do not require specular reflection in order to better observe their surface.

Here is what I have done:

```
auto phongMaterial = vsg::PhongMaterialValue::create();
phongMaterial->value().alphaMask = 1.0f;
phongMaterial->value().alphaMaskCutoff = 0.0025f;
phongMaterial->value().specular = { 0.0,0.0,0.0,1.0 };
```

**Reply by robertosfield** [reply]
Is it all working fine now?

**Reply by mingkangzh** [reply]
The mirror reflection effect has disappeared, that is, the highlighted part has disappeared.

When the camera is facing the point cloud along the light source, there is still only white because the point cloud normal points directly towards the camera.

I changed the normal direction of the point cloud, and when the camera is facing the point cloud, purple appears.

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/e9b17956-3459-4edf-97de-df0303e12300)

I have made the following settings:



        auto& defines = config->shaderHints->defines;
        struct SetPipelineStates : public vsg::Visitor
        {
            bool blending = false;
            SetPipelineStates(bool in_blending) { blending = in_blending; }

            void apply(vsg::Object& object) { object.traverse(*this); }
            void apply(vsg::RasterizationState& rs) override
            {
                rs.cullMode = VK_CULL_MODE_NONE;
            }

            void apply(vsg::InputAssemblyState& ias) { ias.topology = VK_PRIMITIVE_TOPOLOGY_POINT_LIST; }
            void apply(vsg::ColorBlendState& cbs) { cbs.configureAttachments(blending); }
        } sps(blending);

        config->accept(sps);

        defines.insert("VSG_TWO_SIDED_LIGHTING");

        config->init();

The back is still black with some purple. The color did not appear the same as FRONT.

I also wonder why the back is not entirely black

**Reply by mingkangzh** [reply]
Sorry again, everything is normal.

I customized the Switch. The default is not to add light sources to the scene, as in the previous example, both ambient and directional lights were not added to the scene graph.

I changed to adding the light source to the scene by default, and now everything is normal.

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/71ec38c2-656c-4539-809d-328fee2fbb11)

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/87624060/9ae84ac8-1cbc-4af5-8631-51848b6b530b)
